### PR TITLE
feat: add native Ollama /api/chat and /api/generate instrumentation

### LINKED ISSUE #2650: Add native Ollama chat and generation instrumentation
## Parent

- #1854

## Context

OBI does not currently recognize Ollama's native `/api/chat` and `/api/generate` protocols. These endpoints use a different schema from the OpenAI-compatible `/v1/` endpoints: token counts are reported as `prompt_eval_count` and `eval_count`, chat output is under `message`, and generation output is under `response`.

OpenLIT adds a dedicated parser and routing:

- [OpenLIT config patch](https://github.com/openlit/openlit/blob/main/openlit-controller/patches/config.patch)
- [OpenLIT span patch](https://github.com/openlit/openlit/blob/main/openlit-controller/patches/span.patch)
- [OpenLIT HTTP transform patch](https://github.com/openlit/openlit/blob/main/openlit-controller/patches/http_transform.patch)
- [OpenLIT Ollama provider](https://github.com/openlit/openlit/blob/main/openlit-controller/_obi-providers/ollama/ollama.go)

The upstream implementation must correct the operation mapping, content structure, and streaming limitations in that patch.

## Requirements

- [ ] Add explicit `ollama.enabled` payload-extraction configuration and documentation.
- [ ] Restrict detection to HTTP client requests ending in `/api/chat` or `/api/generate`.
- [ ] Keep Ollama native detection separate from its OpenAI-compatible `/v1/` endpoints.
- [ ] Support both non-streaming JSON and streaming newline-delimited JSON. Ollama enables streaming by default.
- [ ] Set `gen_ai.provider.name = "ollama"` consistently on spans and metrics.
- [ ] Map `/api/chat` to `gen_ai.operation.name = "chat"`.
- [ ] Map `/api/generate` to `gen_ai.operation.name = "text_completion"`; do not export the raw endpoint name `generate`.
- [ ] Name spans `{gen_ai.operation.name} {gen_ai.request.model}`.
- [ ] Preserve request and response models separately and report the server-returned response model when available.
- [ ] Extract `prompt_eval_count`, `eval_count`, `done_reason`, streaming state, and provider errors when available.
- [ ] Emit usage attributes and metrics only when the corresponding token count is available.
- [ ] Normalize chat input into the OTel `role` plus `parts` message structure.
- [ ] Normalize generate prompts into a user text part and generated text into an assistant output message.
- [ ] Keep chat-history system messages in `gen_ai.input.messages`; map the separate `system` field of `/api/generate` to `gen_ai.system_instructions`.
- [ ] Keep message content opt-in.
- [ ] Update the configuration schema, support matrix, feature documentation, and examples.

## Tests

- [ ] Non-streaming `/api/chat`.
- [ ] Default streamed `/api/chat`, including final usage and finish reason.
- [ ] Non-streaming `/api/generate`.
- [ ] Default streamed `/api/generate`, including final usage and finish reason.
- [ ] Tool-call and system-message chat payloads.
- [ ] Negative paths, malformed payloads, and unrelated services with similar JSON.
- [ ] Request and response model differences.
- [ ] Input, output, and system-instruction values validate against the selected OTel GenAI schemas.
- [ ] Trace and metric attributes use the same provider and operation values.

## Semantic-convention constraints

`ollama` is not currently a well-known provider enumeration, but custom provider values are permitted. The native generation endpoint is a prompt-to-text completion, so the applicable predefined operation is `text_completion`; when a predefined operation applies, it must be used.

Raw Ollama `{"role":"...","content":"..."}` messages and plain output strings must not be exported directly as `gen_ai.input.messages` or `gen_ai.output.messages`; those attributes must follow the corresponding OTel schemas.

References:

- [Ollama chat API](https://docs.ollama.com/api/chat)
- [Ollama generate API](https://docs.ollama.com/api/generate)
- [GenAI span conventions](https://github.com/open-telemetry/semantic-conventions-genai/blob/main/docs/gen-ai/gen-ai-spans.md)
- [GenAI input message schema](https://github.com/open-telemetry/semantic-conventions-genai/blob/main/model/gen-ai/gen-ai-input-messages.json)
- [GenAI output message schema](https://github.com/open-telemetry/semantic-conventions-genai/blob/main/model/gen-ai/gen-ai-output-messages.json)

### PR DESCRIPTION
**Pull request by NameHaibinZhang**
## Summary

Add native detection for Ollama's `/api/chat` and `/api/generate` endpoints, enabling GenAI observability for Ollama-hosted models without requiring an OpenAI-compatible proxy.

Closes #2650

## Validation

- [x] I have read and followed the [contributing guidelines](https://github.com/open-telemetry/opentelemetry-ebpf-instrumentation/blob/main/CONTRIBUTING.md)
- [x] If this enhances / fixes / changes a core feature, I have updated the [features documentation](https://github.com/open-telemetry/opentelemetry-ebpf-instrumentation/blob/main/devdocs/features.md) and [support matrix](https://github.com/open-telemetry/opentelemetry-ebpf-instrumentation/blob/main/SUPPORT_MATRIX.md) as needed.

### CODE REVIEW

**Review by mariomac — APPROVED** [review]
Looks great to me!

If the code is written from scratch, is ready to merge!

If there is any part of the code taken from https://github.com/openlit/openlit I'd add a crediting comment on the start of the files that contain the comment, linking to the project.

**Comment by NameHaibinZhang** [comment]
The code is written from scratch, based on the existing GenAI detection patterns in this repository (OpenAI, Anthropic, etc.) and the Ollama API documentation. No code was taken from openlit or any other external project.

File: pkg/appolly/app/request/span.go [hunk 1]
```diff
@@ -298,6 +300,7 @@ type GenAI struct {
 	Embedding        *VendorEmbedding
 	Rerank           *VendorRerank
 	Retrieval        *VendorRetrieval
+	Ollama           *VendorOpenAI
```
Comment (mmat11) [inline pkg/appolly/app/request/span.go:303]: I see we have Qwen which reuses the OpenAI struct, we also have OpenAICompatible and now Ollama, should we rename the struct to something less confusing? 
Also, what's OpenAICompatible? Should we merge some of them instead, or have "flavors"?

Comment (NameHaibinZhang) [reply pkg/appolly/app/request/span.go:303]: `VendorOpenAI` describes the response format originated by OpenAI (model, choices, usage with prompt/completion tokens), which has become the de facto standard. Qwen already reuses it for the same reason — the struct name reflects the format's origin, not the consumer.

`OpenAICompatible` is specifically for services that expose the `/v1/chat/completions` endpoint without a dedicated provider identity (e.g. vLLM, LocalAI). Ollama's native `/api/` endpoints use a different URL structure but share the same response semantics, so reusing the struct is natural.

I'd prefer to keep the naming as-is for now to stay consistent with the existing Qwen pattern and keep this PR scoped. Happy to discuss a rename in a separate issue if the team feels strongly about it.

File: pkg/ebpf/common/http/ollama.go [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+	// Map token counts: Ollama uses prompt_eval_count / eval_count.
+	if ollamaResp.PromptEvalCount > 0 {
+		parsed.Usage.InputTokens = ollamaResp.PromptEvalCount
+	}
+	if ollamaResp.EvalCount > 0 {
+		parsed.Usage.OutputTokens = ollamaResp.EvalCount
+	}
+
+	// Build request info.
+	parsed.Request = request.OpenAIInput{
```
Comment (grcevski) [inline pkg/ebpf/common/http/ollama.go:124]: I believe Ollama enables streams [by default](https://docs.ollama.com/api/chat#body-stream), so nil stream should mean true:
We should do something like this:
```
stream := true
if ollamaReq.Stream != nil {
	stream = *ollamaReq.Stream
}
```

File: pkg/ebpf/common/http/ollama.go [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+			_ = json.Unmarshal(ollamaResp.Message, &msg)
+		}
+		if msg.Content != "" || ollamaResp.DoneReason != "" {
+			buildOllamaChoices(parsed, msg.Role, msg.Content, ollamaResp.DoneReason)
+		}
+	} else if ollamaResp.Response != "" || ollamaResp.DoneReason != "" {
+		buildOllamaChoices(parsed, "assistant", ollamaResp.Response, ollamaResp.DoneReason)
+	}
+
+	parsed.ToolCalls = toolCalls
```
Comment (grcevski) [inline pkg/ebpf/common/http/ollama.go:158]: I think we need to add Ollama as a case in getSpanToolCalls in tracesgen.go.

File: pkg/config/payload_extraction.go [hunk 4]
```diff
@@ -98,14 +98,16 @@ type GenAIConfig struct {
 	Rerank RerankConfig `yaml:"rerank"`
 	// Vector retrieval payload extraction and parsing (Pinecone, Qdrant, Milvus, Chroma, Weaviate, etc.)
 	Retrieval RetrievalConfig `yaml:"retrieval"`
+	// Ollama native API payload extraction and parsing
+	Ollama OllamaConfig `yaml:"ollama"`
```
Comment (grcevski) [inline pkg/config/payload_extraction.go:106]: Can we please add Ollama to the config-v2 conversion just like the other GenAI code?

**Review by grcevski — COMMENTED** [review]
Great stuff! I left a few comments.

File: pkg/ebpf/common/http/ollama.go (hunk 2 above)
Comment (NameHaibinZhang) [reply pkg/ebpf/common/http/ollama.go:124]: Fixed. Changed the stream logic to `ollamaReq.Stream == nil || *ollamaReq.Stream` so that an omitted `stream` field defaults to `true` per the [Ollama API docs](https://docs.ollama.com/api/chat#body-stream). Also added a test case `TestOllamaSpan_ChatDefaultStream` to verify the nil-means-true behavior.

File: pkg/ebpf/common/http/ollama.go (hunk 3 above)
Comment (NameHaibinZhang) [reply pkg/ebpf/common/http/ollama.go:158]: Fixed. Added `case span.GenAI.Ollama != nil` to `getSpanToolCalls` in `tracesgen.go` so Ollama tool calls are properly extracted.

File: pkg/config/payload_extraction.go (hunk 4 above)
Comment (NameHaibinZhang) [reply pkg/config/payload_extraction.go:106]: Fixed. Added `"ollama"` to the config-v2 conversion pipeline:
- `export_parity.go`: new `payloadExtractorOllama` constant + export logic
- `import.go`: added to `validV2HTTPPayloadExtractor` and `applyV2HTTPPayloadExtractorMembership`
- `obi-extension.schema.json`: added to the enum
- `default-configuration.yaml`: updated the supported-values comment
- `check-config-v2-parity/main.go`: added to the parity check list
- Updated round-trip tests in both `import_http_test.go` and `export_test.go`

**Review by marctc — APPROVED** [review]
LGTM

**Review by grcevski — APPROVED** [review]
LGTM!
